server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java

```diff
@@ -127,6 +127,28 @@ public String getName() {
         return name;
     }
 
+    /**
+     * Rewrite the sub-aggregations in the buckets in this aggregation.
+     * The default implementation throws an exception because most
+     * aggregations don't <strong>have</strong> buckets in them. It
+     * should be overridden by aggregations that contain buckets.
+     */
+    public InternalAggregation rewriteBuckets(BucketRewriter rewriter) {
```

Comment:
Let's beef up the docs here, it took me a while to really understand what this is doing.

E.g. an implementer of this will iterate over their internal buckets and apply the rewriter to each bucket, generating a modified/rewritten representation of their buckets which is then returned.  

Also I think it's probably important to note that this returns a new version of the InternalAggregations... but the callee remains immutable and does not itself change.

Reply:
Perhaps rename the method to `copyAndRewriteBuckets()` or similar, to help make it clear what's going on?

Reply:
++